sync/optimistic.md

```diff
@@ -0,0 +1,340 @@
+
+A consensus engine MUST be able to retrospectively (i.e., after import) modify
+the status of `SYNCING` blocks to be either `VALID` or `INVALID` based upon responses
+from an execution engine. I.e., perform the following transitions:
+
+- `SYNCING` -> `VALID`
+- `SYNCING` -> `INVALID`
+
+When a block transitions from `SYNCING` -> `VALID`, all *ancestors* of the
+block MUST also transition from `SYNCING` -> `VALID`. Such a block is no longer
+considered "optimistically imported".
+
+When a block transitions from `SYNCING` -> `INVALID`, all *descendants* of the
+block MUST also transition from `SYNCING` -> `INVALID`.
+
+When a block transitions from the `SYNCING` state it is removed from the set of
+`store.optimistic_roots`.
+
+When a "merge block" (i.e. a block which enables execution) is declared to be
+`VALID` by an execution engine (either directly or indirectly), the full
+[`validate_merge_block`][] MUST be run against the merge block. If the block
+fails [`validate_merge_block`][], the merge block MUST be treated the same as
+an `INVALID` block (i.e., it and all its descendants are invalidated and
+removed from the block tree).
+
+### Execution Engine Errors
+
+When an execution engine returns an error or fails to respond to a payload
+validity request for some block, a consensus engine:
+
+- MUST NOT optimistically import the block.
+- MUST NOT apply the block to the fork choice store.
+- MAY queue the block for later processing.
+
+### Assumptions about Execution Engine Behaviour
+
+This specification assumes execution engines will only return `SYNCING` when
+there is insufficient information available to make a `VALID` or `INVALID`
+determination on the given `ExecutionPayload` (e.g., the parent payload is
+unknown). Specifically, `SYNCING` responses should be fork-specific, in that
+the search for a block on one chain MUST NOT trigger a `SYNCING` response for
+another chain.
+
+### Re-Orgs
+
+The consensus engine MUST support any chain reorganisation which does *not*
+affect the justified checkpoint. The consensus engine MAY support re-orgs
+beyond the justified checkpoint.
+
+If the justified checkpoint transitions from `SYNCING` -> `INVALID`, a
+consensus engine MAY choose to alert the user and force the application to
+exit.
+
+## Fork Choice
+
+Consensus engines MUST support removing blocks from fork choice that transition
+from `SYNCING` to `INVALID`. Specifically, a block deemed `INVALID` at any
+point MUST NOT be included in the canonical chain and the weights from those
+`INVALID` blocks MUST NOT be applied to any `VALID` or `SYNCING` ancestors.
+
+### Fork Choice Poisoning
+
+During the merge transition it is possible for an attacker to craft a
+`BeaconBlock` with an execution payload that references an
+eternally-unavailable `body.execution_payload.parent_hash` (i.e., the parent
+hash is random bytes). In rare circumstances, it is possible that an attacker
```

Comment:
Is "rare" the proper word here? Is it rare from a probability standpoint or actually difficult or costly to induce from a technical or crypto-economic standpoint?

Reply:
Ah, this is "rare" as in it's a blck that *triggers* a justification and thus causes issues with `filter_block_tree`

Am I correct in assessing that a "fix" to the issue found a couple of weeks ago would sidestep this deadlock problem @paulhauner ?

Reply:
> Is "rare" the proper word here? Is it rare from a probability standpoint or actually difficult or costly to induce from a technical or crypto-economic standpoint?

I was using rare since the conditions required for this bug rarely occur on mainnet. I was being somewhat vague to avoid needing to disclose too much about the attack. I'm open to re-wording, though!

> Am I correct in assessing that a "fix" to the issue found a couple of weeks ago would sidestep this deadlock problem @paulhauner ?

I've also been thinking about this. I'd need to see the concrete proposal, but I don't think it's the silver bullet here. Even if fork choice can consider the *potential* just/fin of a block (i.e., the "fix"), I think we're still open to the scenario where one chain contains more attestations and can therefore justify in the next epoch whilst the other chain can't.

Reply:
Assuming adversary power is 1/3 and 32 honest proposals is enough to justify a checkpoint, the probability of *not* justifying a checkpoint within 4 epochs is `1.11**-16` according to binomial distribution.

In our case it means that should an adversary be so lucky to get a chance to poison the fork choice the chance of successful attack with 4 epochs timeout is around 0.

1/3 power here makes a lot of sense as this is a liveness attack and Casper FFG has 1/3 honest assumption to maintain liveness. I am not feeling that strong about 32 honest proposals, but it seems that this is more than enough in a good networking conditions as the mainnet always justifies in an epoch.

So, I would say that we have a pretty strong reasoning behind this fix and its parameter.

Also, worth noting that an optimistic node justifying epoch `N` by applying an honest chain *prior to* applying malicious chain which justifies the same epoch `N` is enough to prevent poisoning. Justifying a higher epoch by an honest chain is not necessarily needed.

Reply:
Got it.

concrete proposal coming soon -- cc @adiasg